util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+			}
 		}
 
-	case IsValueMap(ni.FieldValue):
-		for _, key := range ni.FieldValue.MapKeys() {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.MapIndex(key)
-			nn.FieldKey = key
-			nn.FieldKeys = ni.FieldValue.MapKeys()
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeMap(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		if IsNilOrInvalidValue(v) {
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: []string{schema.Name},
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for _, key := range ni.FieldValue.MapKeys() {
+				nn := *ni
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = []string{schema.Name}
+				nn.FieldValue = ni.FieldValue.MapIndex(key)
+				nn.FieldKey = key
+				nn.FieldKeys = ni.FieldValue.MapKeys()
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 	}
 
-	return nil
+	return errs
+}
+
+// GetNodes returns the nodes in the data tree at the indicated path, relative
+// to the supplied root. If for a list node is missing, all values in the list
+// are returned. If the key is partial, all nodes matching the values present
+// in the key are returned.
+// If the root is the tree root, the path may be absolute.
+// GetNodes returns an error if the path is not found in the tree, or an element
+// along the path is nil.
+func GetNodes(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	return getNodesInternal(schema, root, path)
+}
+
+// getNodesInternal is the internal implementation of GetNode. In addition to
+// GetNode functionality, it can accept non GoStruct types e.g. map for a keyed
+// list, or a leaf.
+func getNodesInternal(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	if IsValueNil(root) {
+		ResetIndent()
+		return nil, nil, nil
+	}
+	if len(path.GetElem()) == 0 {
+		ResetIndent()
+		return []interface{}{root}, []*yang.Entry{schema}, nil
+	}
+	if schema == nil {
+		return nil, nil, fmt.Errorf("nil schema for data element type %T, remaining path %v", root, path)
+	}
+	// Strip off the absolute path prefix since the relative and absolute paths
+	// are assumed to be equal.
+	if path.GetElem()[0].GetName() == "" {
+		path.Elem = path.GetElem()[1:]
+	}
+
+	Indent()
+	DbgPrint("GetNode next path %v, value %v", path.GetElem()[0], ValueStr(root))
+
+	switch {
+	case schema.IsContainer() || (schema.IsList() && IsTypeStructPtr(reflect.TypeOf(root))):
+		// Either a container or list schema with struct data node (which could
+		// be an element of a list).
+		return getNodesContainer(schema, root, path)
+	case schema.IsList():
+		// A list schema with the list data node. Must find the element selected
+		// by the path.
```

Comment:
By the keys supplied to the list, I think we always specify the node with the path.

Reply:
I think this comment was confusing. Just trying to say that the function we are calling still needs to find the list element we want, using the supplied path (with keys).